tools/testbench/topology_ipc4.c

```diff
@@ -0,0 +1,1304 @@
+			strerror(errno));
+		return -errno;
+	}
+
+	/* file size */
+	if (fseek(file, 0, SEEK_END)) {
+		fprintf(stderr, "error: can't seek to end of topology: %s\n",
+			strerror(errno));
+		fclose(file);
+		return -errno;
+	}
+	ctx->tplg_size = ftell(file);
+	if (fseek(file, 0, SEEK_SET)) {
+		fprintf(stderr, "error: can't seek to beginning of topology: %s\n",
+			strerror(errno));
+		fclose(file);
+		return -errno;
+	}
+
+	/* load whole topology into memory */
+	ctx->tplg_base = calloc(ctx->tplg_size, 1);
+	if (!ctx->tplg_base) {
+		fprintf(stderr, "error: can't alloc buffer for topology %zu bytes\n",
+			ctx->tplg_size);
+		fclose(file);
+		return -ENOMEM;
+	}
+	ret = fread(ctx->tplg_base, ctx->tplg_size, 1, file);
+	if (ret != 1) {
+		fprintf(stderr, "error: can't read topology: %s\n", strerror(errno));
+		free(ctx->tplg_base);
+		fclose(file);
+		return -errno;
+	}
+	fclose(file);
+
+	/* initialize widget, route, pipeline and pcm lists */
+	list_init(&tp->widget_list);
+	list_init(&tp->route_list);
+	list_init(&tp->pcm_list);
+	list_init(&tp->pipeline_list);
+
+	while (ctx->tplg_offset < ctx->tplg_size) {
+		/* read next topology header */
+		hdr = tplg_get_hdr(ctx);
+
+		tplg_debug("type: %x, size: 0x%x count: %d index: %d\n",
+			   hdr->type, hdr->payload_size, hdr->count, hdr->index);
+
+		ctx->hdr = hdr;
+
+		/* parse header and load the next block based on type */
+		switch (hdr->type) {
+		/* load dapm widget */
+		case SND_SOC_TPLG_TYPE_DAPM_WIDGET:
+
+			tplg_debug("number of DAPM widgets %d\n", hdr->count);
+
+			/* update max pipeline_id */
+			ctx->pipeline_id = hdr->index;
+
+			for (i = 0; i < hdr->count; i++) {
+				ret = tb_load_widget(tp);
+				if (ret < 0) {
+					fprintf(stderr, "error: loading widget\n");
+					return ret;
+				}
+				ctx->comp_id++;
+			}
+			break;
+
+		/* set up component connections from pipeline graph */
+		case SND_SOC_TPLG_TYPE_DAPM_GRAPH:
+			if (tb_register_graph(tp, hdr->count) < 0) {
+				fprintf(stderr, "error: pipeline graph\n");
+				ret = -EINVAL;
+				goto out;
+			}
+			break;
+
+		case SND_SOC_TPLG_TYPE_PCM:
+			ret = tb_parse_pcm(tp, hdr->count);
+			if (ret < 0)
+				goto out;
+			break;
+
+		default:
+			tplg_skip_hdr_payload(ctx);
+			break;
+		}
+	}
+
+	/* assign pipeline to every widget in the widget list */
+	list_for_item(item, &tp->widget_list) {
+		struct tplg_comp_info *comp_info = container_of(item, struct tplg_comp_info, item);
+		struct list_item *pipe_item;
+
+		list_for_item(pipe_item, &tp->pipeline_list) {
+			struct tplg_pipeline_info *pipe_info;
+
+			pipe_info = container_of(pipe_item, struct tplg_pipeline_info, item);
+			if (pipe_info->id == comp_info->pipeline_id) {
+				comp_info->pipe_info = pipe_info;
+				break;
+			}
+		}
+
+		if (!comp_info->pipe_info) {
+			fprintf(stderr, "warning: failed  assigning pipeline for %s\n",
+				comp_info->name);
+		}
+	}
+
+out:
+	/* free all data */
+
+	// TODO: Check this
```

Comment:
Hmm, maybe a bit more verbose if we leave this to codebase as a TODO..?

Reply:
I think this is remaining from some old code change try and I've forgot to clean it up. It's clear that ctx->tplg_base should be freed. It was was also missing from above load widget.